integration-cli/docker_api_attach_test.go

```diff
@@ -99,128 +100,81 @@ func (s *DockerSuite) TestGetContainersWsAttachContainerNotFound(c *check.C) {
 
 func (s *DockerSuite) TestPostContainersAttach(c *check.C) {
 	testRequires(c, DaemonIsLinux)
-	out, _ := dockerCmd(c, "run", "-dit", "busybox", "cat")
-
-	r, w := io.Pipe()
-	defer r.Close()
-	defer w.Close()
-
-	conn, err := sockConn(time.Duration(10 * time.Second))
-	c.Assert(err, check.IsNil)
-
-	containerID := strings.TrimSpace(out)
 
-	req, err := http.NewRequest("POST", "/containers/"+containerID+"/attach?stream=1&stdin=1&stdout=1&stderr=1", bytes.NewReader([]byte{}))
-	c.Assert(err, check.IsNil)
-
-	client := httputil.NewClientConn(conn, nil)
-	defer client.Close()
-
-	// Do POST attach request
-	resp, err := client.Do(req)
-	c.Assert(resp.StatusCode, check.Equals, http.StatusOK)
-	// If we check the err, we get a ErrPersistEOF = &http.ProtocolError{ErrorString: "persistent connection closed"}
-	// This means that the remote requested this be the last request serviced, is this okay?
-
-	// Test read and write to the attached container
-	expected := []byte("hello")
-	actual := make([]byte, len(expected))
-
-	outChan := make(chan error)
-	go func() {
-		_, err := r.Read(actual)
-		outChan <- err
-		close(outChan)
-	}()
-
-	inChan := make(chan error)
-	go func() {
-		_, err := w.Write(expected)
-		inChan <- err
-		close(inChan)
-	}()
+	expectSuccess := func(conn net.Conn, br *bufio.Reader, stream string, tty bool) {
+		defer conn.Close()
+		expected := []byte("success")
+		_, err := conn.Write(expected)
+		c.Assert(err, check.IsNil)
 
-	select {
-	case err := <-inChan:
+		conn.SetReadDeadline(time.Now().Add(time.Second))
+		lenHeader := 0
+		if !tty {
+			lenHeader = 8
+		}
+		actual := make([]byte, len(expected)+lenHeader)
+		_, err = io.ReadFull(br, actual)
 		c.Assert(err, check.IsNil)
-	case <-time.After(5 * time.Second):
-		c.Fatal("Timeout writing to stdout")
+		if !tty {
+			fdMap := map[string]byte{
+				"stdin":  0,
+				"stdout": 1,
+				"stderr": 2,
+			}
+			c.Assert(actual[0], check.Equals, fdMap[stream])
+		}
+		c.Assert(actual[lenHeader:], check.DeepEquals, expected, check.Commentf("Attach didn't return the expected data from %s", stream))
 	}
 
-	select {
-	case err := <-outChan:
+	expectTimeout := func(conn net.Conn, br *bufio.Reader, stream string) {
+		defer conn.Close()
+		_, err := conn.Write([]byte{'t'})
 		c.Assert(err, check.IsNil)
-	case <-time.After(5 * time.Second):
-		c.Fatal("Timeout reading from stdin")
-	}
 
-	if !bytes.Equal(expected, actual) {
```

Comment:
As long as there are no bugs in the golang stdlib and the kernel this will almost always be true.